mangaki/mangaki/utils/knn.py

```diff
@@ -118,20 +87,29 @@ def fit(self, X=None, y=None, whole_dataset=False):
             self.M[user_id, work_id] = rating
         for work_id in self.nb_ratings:
             self.mean_score[work_id] = self.sum_ratings[work_id] / self.nb_ratings[work_id]
-        self.get_all_neighbors()
 
     def predict(self, X):
+        self.get_neighbors(list(set(X[:, 0])))  # Compute only relevant neighbors
         y = []
         for my_user_id, work_id in X:
-            """if not my_user_id in self.closest_neighbors:
-                self.get_neighbors(my_user_id)"""
             weight = 0
             predicted_rating = 0
             for user_id in self.closest_neighbors[my_user_id]:
                 their_sim_score = self.closest_neighbors[my_user_id][user_id]
-                their_rating = self.ratings[user_id].get(work_id, self.mean_score.get(work_id, 0))  # Double fallback
-                predicted_rating += their_sim_score * their_rating
-                weight += their_sim_score
+                if self.missing_is_mean:
+                    their_rating = self.ratings[user_id].get(work_id, self.mean_score.get(work_id, 0))  # Double fallback
```

Comment:
Under which context can the double fallback happens? Why would the `mean_score` dict be empty? (maybe a comment about it).